src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,580 @@
+            packer.packArrayHeader(streamIds.size());
+            for (String s: streamIds) {
+                packer.packString(s);
+            }
+        } else {
+            packer.packArrayHeader(0);
+        }
+    }
+    
+    private Object[] bindArguments(MessageUnpacker unpacker, List<Class<?>> paramTypes) throws IOException {
+        int argumentCount = unpacker.unpackArrayHeader();
+        
+        if (paramTypes.size() != argumentCount) {
+            throw new RuntimeException(String.format("Invocation provides %d argument(s) but target expects %d.", argumentCount, paramTypes.size()));
+        }
+        
+        Object[] arguments = new Object[argumentCount];
+        
+        for (int i = 0; i < argumentCount; i++) {
+            arguments[i] = readValue(unpacker, paramTypes.get(i));
+        }
+        
+        return arguments;
+    }
+    
+    private Object readValue(MessageUnpacker unpacker, Class<?> itemType) throws IOException {
+        // This shouldn't ever get called with itemType == null, but we return anyways to avoid NullPointerExceptions
+        if (itemType == null) {
+            return null;
+        }
+        MessageFormat messageFormat = unpacker.getNextFormat();
+        ValueType valueType = messageFormat.getValueType();
+        int length;
+        Object item = null;
+         
+        switch(valueType) {
+            case NIL:
+                unpacker.unpackNil();
+                item = null;
+                break;
+            case BOOLEAN:
+                item = unpacker.unpackBoolean();
+                break;
+            case INTEGER:
+                switch (messageFormat) {
+                case UINT64:
+                    item = unpacker.unpackBigInteger();
+                    break;
+                case INT64:
+                case UINT32:
+                    item = unpacker.unpackLong();
+                    break;
+                default:
+                    item = unpacker.unpackInt();
+                    break;
+                }
+                break;
+            case FLOAT:
+                item = unpacker.unpackDouble();
+                break;
+            case STRING:
+                item = unpacker.unpackString();
+                break;
+            case BINARY:
+                length = unpacker.unpackBinaryHeader();
+                byte[] binaryValue = new byte[length];
+                unpacker.readPayload(binaryValue);
+                item = binaryValue;
+                break;
+            case ARRAY:
+                length = unpacker.unpackArrayHeader();
+                Object[] arrayValue = new Object[length];
+                for (int i = 0; i < length; i++) {
+                    arrayValue[i] = readValue(unpacker, new Object().getClass());
+                }
+                item = arrayValue;
+                //If the itemType is an array, we return an array. Else we convert the array to a list.
+                if (!itemType.isArray()) {
```

Comment:
Is this necessary? I make the distinction between returning a List<T> and an Array. I could probably choose to just return all collections as Arrays, and let the server figure out what to do.

Reply:
Does the JsonHubProtocol always return an ArrayList? We should keep the behavior as close to JsonHubProtocol as we can.

Reply:
It looks like it returns whatever the specified type was, so it does make the distinction between Arrays and Lists (as per gson): https://www.javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/Gson.html